src/sync.rs

```diff
@@ -1,11 +1,44 @@
-use config::Config;
-use ignore::Ignore;
 use std::path::Path;
 use std::path::PathBuf;
 use std::process::Command;
+use std::sync::mpsc::TryRecvError::*;
+use std::thread;
+use std::time::{Duration, Instant};
+
+use bus::BusReader;
+use crossbeam_channel::Receiver;
+use crossbeam_channel::Sender;
+use crossbeam_channel::unbounded;
+
+use config::Config;
+use ignore::Ignore;
+use remote_command::RemoteCommandResult;
+
+#[derive(Debug, PartialEq, Clone)]
+pub enum PushResult {
```

Comment:
Any reason not to make this a [type synonym](https://doc.rust-lang.org/book/ch19-04-advanced-types.html#creating-type-synonyms-with-type-aliases) instead so you can reuse all the functionality of `Result` directly?

Reply:
Or actually, how about something like:
```suggestion
pub struct PushResult {
    pub duration: Duration,
    pub result: Result<(), String>,
}
```
I think this would reflect the intent more accurately, looking at `fn push()` below.

Reply:
The only reason to not make a synonym was to better support extension with additional fields, but I think best version would be to combine custom structs with raw Result

```rust
pub struct PushOk {
  pub duration: Duration
}

pub struct PushErr {
  pub duration: Duration
}

Result<PushOk, PushErr>
```

I def feel weird that enums can't have named fields, structs seem to provide more protection from accidental value mismatch

Reply:
I guess you meant... `PushEr`...

Reply:
IMO it would still be better to have a product type at the top since you'll always want the duration, either via a named struct as I suggested above or an (anonymous) tuple. This also lets you punt on custom structs for the Ok/Err cases until you have extra values there - the compiler will enable safe refactoring later. This lets you avoid adding too much complexity; e.g., the `execute_rsync(&mut command)` on master which is currently in your PR:
```
  match execute_rsync(&mut command) {
        Err(reason) => Err(PushErr {
            duration: start_time.elapsed(),
            message: reason,
        }),
        Ok(_) => Ok(PushOk {
            duration: start_time.elapsed()
        }),
    }
```
could be just
```
  (start_time.elapsed(), execute_rsync(&mut command))
```

Similarly, code like
```
let remote_command_duration = match remote_command_result {
    Err(err) => err.duration,
    Ok(ok) => ok.duration
};
``` goes away.

Reply:
I'd like a product top-level type but I'd also like to immediately be able to distinguish its subtype (if needed)

In Kotlin (best example I have on mind) I'd do this:

```kotlin
sealed class PushResult(val duration: Duration) {
  data class Ok(override val duration: Duration): PushResult(duration)
  data class Err(override val duration: Duration): PushResult(duration)
}
```

This way it's still an enum (sealed class in Kotlin) while also containing always-present data in each case.

In Rust I guess I could do something like this:

```rust
pub enum PushResult {
    Ok(Duration),
    Err(Duration)
}

pub trait _PushResult {
    fn duration(&self) -> &Duration;
}

pub impl _PushResult for PushResult {
    fn duration(&self) -> &Duration {
        match self {
            PushResult::Err(duration) => duration,
            PushResult::Ok(duration) => duration
        }
    }
}
```

wdyt?